File: src/main/stack/StackWindow.java
package main.stack;

import main.interfaces.DefaultWindowsInterface;
import main.interfaces.GridInterface;
import main.interfaces.LinkedListInterface;
import main.interfaces.StackLightWeightInterface;
import shapes.MyArrow;
import utils.AnimationHelper;

import javax.swing.*;
import java.awt.*;
import java.util.Objects;

public class StackWindow extends JPanel implements StackLightWeightInterface, GridInterface {

    private final int myWidth = StackLightWeightInterface.width; // Interface constants are static final
    private final int myHeight = StackLightWeightInterface.height;
    private final MyArrow myArrow = new MyArrow(80, 12);
    private LinkedListInterface.VisualNode top;
    private int size = 1;
    private final int nodeHeight = 80;
    private int nodeWidth = 100;
    private final int startX = returnClosest(myWidth / 5, myWidth / 3, SPACING + 5);
    private final int endX = returnClosest(myWidth - 2 * nodeWidth, myWidth - nodeWidth, SPACING + 5);
    private int startY = nodeHeight * 2;
    private int endY = returnClosest(myHeight - 4 * nodeHeight, myHeight - 3 * nodeHeight, SPACING + 5);
    private int dynamicHeight = startY;
    private boolean stopPop = false;
    private boolean isAnimating = false;
    private final int animationSpeed = 15;

    public StackWindow() {
        nodeWidth = endX - startX - 60;
        top = new LinkedListInterface.VisualNode(0, startX + 30, endY - nodeHeight - SPACING);
        setBackground(new Color(0xA0F29));
        setPreferredSize(new Dimension(myWidth, myHeight));
    }

    @Override
    protected void paintComponent(Graphics g1) {
        super.paintComponent(g1);
        Graphics2D g = (Graphics2D) g1;
        g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        g.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY);
        drawGrid(g, new Color(0x1C233D));
        drawTitle(g);
        drawBasket(g);
        LinkedListInterface.VisualNode temp = top;
        while (temp != null) {
            drawNode(g, temp);
            temp = temp.getNextNode();
        }
    }

    private void drawTitle(Graphics2D g) {
        g.setColor(Color.WHITE);
        g.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 28));
        g.drawString("Stack Visualization (LIFO)", myWidth / 2 - 170, 40);

        g.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 18));
        g.setColor(new Color(0xFFD700));
        g.drawString("Size: " + size, myWidth / 2 - 40, 70);
    }

    public void setCamCentered(JScrollPane scrollPane) {
        if (isAnimating)
            return;

        SwingUtilities.invokeLater(() -> {
            JViewport viewport = scrollPane.getViewport();
            Rectangle viewRect = viewport.getViewRect();
            int currentX = viewRect.x;
            int contentHeight = getPreferredSize().height;
            int viewportHeight = viewRect.height;
            int targetCenterY = top.yPos + nodeHeight / 2;
            int unclampedY = targetCenterY - viewportHeight / 2;
            int maxY = Math.max(0, contentHeight - viewportHeight);
            int clampedY = Math.max(0, Math.min(unclampedY, maxY));

            int startY = viewRect.y;
            int distance = clampedY - startY;
            int steps = 10;

            // Java Timer is different from Kotlin Timer. Using Swing Timer.
            new Timer(20, new java.awt.event.ActionListener() {
                int currentStep = 0;

                @Override
                public void actionPerformed(java.awt.event.ActionEvent e) {
                    currentStep++;
                    float progress = AnimationHelper.easeInOut((float) currentStep / steps);
                    int newY = startY + (int) (distance * progress);
                    viewport.setViewPosition(new Point(currentX, newY));

                    if (currentStep >= steps) {
                        ((Timer) e.getSource()).stop();
                        viewport.setViewPosition(new Point(currentX, clampedY));
                    }
                }
            }).start();
        });
    }

    @Override
    public void drawGrid(Graphics2D g, Color color) {
        Color retColor = g.getColor();
        g.setColor(color);
        for (int i = 0; i < DefaultWindowsInterface.width; i += SPACING) {
            g.drawLine(i, 0, i, getPreferredSize().height);
        }
        for (int i = 0; i < getPreferredSize().height; i += SPACING) {
            g.drawLine(0, i, DefaultWindowsInterface.width, i);
        }
        g.setColor(retColor);
    }

    private void shiftElements() {
        int shiftY = nodeHeight * 4;
        LinkedListInterface.VisualNode temp = top;
        while (temp != null) {
            temp.yPos += shiftY;
            temp = temp.getNextNode();
        }
    }

    private void resizePush() {
        if (top.yPos % myHeight < 20) {
            endY += nodeHeight * 4;
            shiftElements();
        }
        setPreferredSize(new Dimension(myWidth, endY + nodeHeight * 2));
        revalidate();
        repaint();
    }

    @Override
    public void push(int value) {
        if (isAnimating)
            return;

        size++;
        resizePush();
        stopPop = false;

        LinkedListInterface.VisualNode newNode = new LinkedListInterface.VisualNode(value, top.xPos, -nodeHeight);
        newNode.setNextNode(top);
        newNode.setNextAddress(top.getAddress());
        LinkedListInterface.VisualNode oldTop = top;
        top = newNode;

        animatePush(newNode, oldTop.yPos - nodeHeight - 10);
    }

    private void animatePush(LinkedListInterface.VisualNode node, int targetY) {
        isAnimating = true;
        new Timer(animationSpeed, e -> {
            if (node.yPos < targetY) {
                node.yPos += 12;
                repaint();
            } else {
                node.yPos = targetY;
                isAnimating = false;
                ((Timer) e.getSource()).stop();
                repaint();
            }
        }).start();
    }

    @Override
    public int pop() {
        if (isAnimating)
            return top.data;

        if (stopPop) {
            dynamicHeight = myHeight;
            setPreferredSize(new Dimension(myWidth, myHeight));
            scrollRectToVisible(new Rectangle(getPreferredSize()));
            repaint();
            return top.data;
        }

        if (sizeSt() == 1 && !stopPop) {
            stopPop = true;
            dynamicHeight -= nodeHeight * 2;
            animatePop(top, -nodeHeight * 3, true);
            endY -= nodeHeight * 2;
            return top.data;
        }

        int retData = top.data;
        LinkedListInterface.VisualNode nodeToRemove = top;
        top.setNextAddress(null);
        LinkedListInterface.VisualNode temp = top.getNextNode();
        top.setNextNode(null);
        top = temp;
        size--;

        animatePop(nodeToRemove, -nodeHeight * 2, false);

        return retData;
    }

    private void animatePop(LinkedListInterface.VisualNode node, int targetY, boolean isSingleElement) {
        isAnimating = true;
        new Timer(animationSpeed, e -> {
            if (node.yPos > targetY) {
                node.yPos -= 12;
                repaint();
            } else {
                node.yPos = targetY;
                isAnimating = false;
                ((Timer) e.getSource()).stop();
                if (!isSingleElement) {
                    resizePop();
                }
                repaint();
            }
        }).start();
    }

    private void resizePop() {
        if (getPreferredSize().height % top.yPos > 100 && sizeSt() > 1) {
            dynamicHeight = top.yPos - nodeHeight * 2;
        }
        setPreferredSize(new Dimension(myWidth, endY));
        scrollRectToVisible(new Rectangle(getPreferredSize()));
        revalidate();
        repaint();
    }

    public int peek() {
        return top.data;
    }

    @Override
    public int sizeSt() {
        return size;
    }

    @Override
    public void drawNode(Graphics2D g, LinkedListInterface.VisualNode node) {
        Color oldColor = g.getColor();
        Stroke resetStroke = g.getStroke();

        // Draw shadow for depth
        g.setColor(new Color(0, 0, 0, 50));
        g.fillRoundRect(node.xPos + 4, node.yPos + 4, nodeWidth, nodeHeight, 10, 10);

        // Draw main node with rounded corners
        g.setColor(new Color(0x1E3A8A));
        g.fillRoundRect(node.xPos, node.yPos, nodeWidth, nodeHeight, 10, 10);

        // Draw border
        g.setColor(new Color(0xFFD700));
        g.setStroke(new BasicStroke(4f));
        g.drawRoundRect(node.xPos, node.yPos, nodeWidth, nodeHeight, 10, 10);

        // Draw vertical divider
        g.setStroke(new BasicStroke(3f));
        g.drawLine(node.xPos + nodeWidth / 2, node.yPos + 5, node.xPos + nodeWidth / 2, node.yPos + nodeHeight - 5);

        // Draw data section
        g.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 26));
        g.setColor(Color.WHITE);
        String dataStr = String.valueOf(node.data);
        FontMetrics fm = g.getFontMetrics(g.getFont());
        g.drawString(dataStr, node.xPos + nodeWidth / 4 - fm.stringWidth(dataStr) / 2, node.yPos + nodeHeight / 2 + 10);

        // Draw label
        g.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 12));
        g.setColor(new Color(0xFFD700));
        g.drawString("Data", node.xPos + nodeWidth / 4 - 15, node.yPos + 20);

        // Draw next address
        g.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 14));
        g.setColor(Color.WHITE);
        String nextAddr = (node.getNextAddress() == null) ? "NULL"
                : node.getNextAddress().substring(0, Math.min(6, node.getNextAddress().length()));
        g.drawString(nextAddr, node.xPos + 3 * nodeWidth / 5 - 10, node.yPos + nodeHeight / 2 + 10);

        g.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 12));
        g.setColor(new Color(0xFFD700));
        g.drawString("Next", node.xPos + 3 * nodeWidth / 5 - 5, node.yPos + 20);

        // Draw address pointer
        g.setColor(Color.WHITE);
        g.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 11));
        String shortAddr = node.getAddress().substring(0, Math.min(8, node.getAddress().length()));
        g.drawString(shortAddr, 30, node.yPos + nodeHeight / 2 + 5);

        myArrow.draw(
                g,
                g.getFontMetrics(g.getFont()).stringWidth(shortAddr) + 35,
                node.yPos + nodeHeight / 2,
                new Color(0xFFD700));

        g.setColor(oldColor);
        g.setStroke(resetStroke);
    }

    private int returnClosest(int startPoint, int endPoint, int divisor) {
        for (int i = startPoint; i <= endPoint; i++) {
            if (i % divisor == 0) {
                return i;
            }
        }
        return startPoint;
    }

    @Override
    public void drawBasket(Graphics2D g) {
        Stroke stroke = g.getStroke();
        Color color = g.getColor();
        g.setStroke(new BasicStroke(15f));
        g.setColor(DefaultWindowsInterface.backgroundColor);
        g.drawLine(startX, dynamicHeight, startX, endY);
        g.drawLine(startX, endY, endX, endY);
        g.drawLine(endX, endY, endX, dynamicHeight);
        g.setStroke(stroke);
        g.setColor(color);
    }
}

File: src/Windows/StackWindow.kt
package Windows
//TODO SCROLLING MECHANISMS
import Windows.Interfaces.DefaultWindowsInterface
import Windows.Interfaces.GridInterface
import Windows.Interfaces.GridInterface.SPACING
import Windows.Interfaces.LinkedListInterface.VisualNode
import Windows.Interfaces.StackLightWeightInterface
import java.awt.*
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import javax.swing.*

class StackWindow : JPanel(), StackLightWeightInterface, GridInterface {
    private val myWidth = StackLightWeightInterface.width;
    private val myHeight = StackLightWeightInterface.height;
    private var top: VisualNode
    private var size = 1
    private val nodeHeight = 80
    private var nodeWidth = 100
    private val startX = returnClosest(myWidth / 5, myWidth / 3, SPACING + 5)
    private val endX = returnClosest(myWidth - 2 * nodeWidth, myWidth - nodeWidth, SPACING + 5)
    private val startY = nodeHeight * 2
    private val endY = returnClosest(myHeight - 4*nodeHeight, myHeight-3*nodeHeight, SPACING + 5)
    private var dynamicHeight=startY
    init {
        nodeWidth = endX - startX - 60
        top = VisualNode(0, startX + 30, endY- nodeHeight-SPACING)
        background = Color(0xA0F29)
        preferredSize = Dimension(myWidth, myHeight)
    }

    protected override fun paintComponent(g1: Graphics) {
        super.paintComponent(g1);
        val g: Graphics2D = g1 as Graphics2D
        drawGrid(g)
        drawBasket(g)
        var temp: VisualNode = top
        resize()
        while (temp != null) {
            drawNode(g, temp)
            temp = temp.nextNode
        }
    }

    private fun resize() {
        if (top.yPos < dynamicHeight) {
            dynamicHeight -= 8 * nodeHeight
            // Update preferred size so JScrollPane knows content grew
            preferredSize = Dimension(myWidth, dynamicHeight + 2 * nodeHeight)
            revalidate()
            repaint()
        }
    }

    fun setCamCentered(scrollPane: JScrollPane) {
        SwingUtilities.invokeLater {
            val viewport = scrollPane.viewport
            // Align viewport so the newest "top" node is visible
            val y = top.yPos.coerceAtLeast(0)
            val x = 0
            viewport.viewPosition = Point(x, y)
        }
    }

    override fun push(value: Int) {
        val newNode = VisualNode(value, top.xPos, top.yPos - nodeHeight - 10)
        newNode.nextNode = top;
        newNode.nextAddress = top.address
        top = newNode
        size++
        repaint()
    }

    override fun pop(): Int {
        if (sizeSt() == 1) {
            return top.data
        }
        val retData = top.data
        top.nextAddress = null
        val temp = top.nextNode
        top.nextNode = null
        //TODO: POSITION CHANGE
        top = temp
        size--
        repaint()
        return retData
    }

    override fun peek(): Int {
        return top.data
    }

    override fun sizeSt(): Int {
        return size;
    }

    override fun drawGrid(g: Graphics2D) {
        g.color = Color(0x1C233D)
        for (i in 0..myWidth step SPACING + 5) {
            g.drawLine(i, 0, i, myHeight)
        }
        for (i in 0..myHeight step SPACING + 5) {
            g.drawLine(0, i, myWidth, i)
        }
    }

    override fun drawNode(g: Graphics2D, node: VisualNode) {
        val oldColor = g.color
        val resetStroke = g.stroke
        g.also {
            it.color = Color(0x1E3A8A)
            it.fillRect(node.xPos, node.yPos, nodeWidth, nodeHeight)
            it.color = Color(0xFFD700)
            it.font = Font(Font.SANS_SERIF, Font.BOLD, 30)
            it.stroke = BasicStroke(6f)
            it.drawString("Data: ${node.data}", node.xPos + nodeWidth / 6, node.yPos + nodeHeight / 2 + 10)
            it.drawLine(node.xPos + nodeWidth / 2, node.yPos + 5, node.xPos + nodeWidth / 2, node.yPos + nodeHeight - 5)
            it.drawString(
                "Next Address: ${
                    when (node.nextAddress) {
                        null -> "NULL"
                        else -> {
                            node.nextAddress
                        }
                    }
                }",
                node.xPos + 9 * nodeWidth / 15,
                node.yPos + nodeHeight / 2 + 10
            )
            it.color = Color.WHITE
            it.font = it.font.deriveFont(10)
            it.drawString("${node.address}  ----->", 50, node.yPos + nodeHeight / 2 + 10)
            it.color = oldColor
            it.stroke = resetStroke
        }
    }

    private fun returnClosest(startPoint: Int, endPoint: Int, divisor: Int): Int {
        for (i in startPoint..endPoint) {
            if (i % divisor == 0) {
                return i
            }
        }
        return startPoint
    }

    override fun drawBasket(g: Graphics2D) {
        val stroke = g.stroke
        val color = g.color
        g.stroke = BasicStroke(15f)
        g.color = DefaultWindowsInterface.themeColorBG
        g.drawLine(startX, dynamicHeight, startX, endY)
        g.drawLine(startX, endY, endX, endY)
        g.drawLine(endX, endY, endX, dynamicHeight)
        g.stroke = stroke
        g.color = color
    }
}

class StackWindowUsable : JPanel(), DefaultWindowsInterface {
    private val myHeight = DefaultWindowsInterface.height;
    private val myWidth = DefaultWindowsInterface.width
    private var textField: JTextField
    private var pushButton: JButton
    private var popButton: JButton
    private val font = Font(Font.SANS_SERIF, Font.BOLD, 20)
    private val visualStackWindow = StackWindow().apply {
        preferredSize= Dimension(myWidth,myHeight)
    }
    init {
        layout = BorderLayout()
        val layeredPane = JLayeredPane().apply {
            preferredSize = Dimension(myWidth, myHeight)
        }
        val scrollPane = JScrollPane(
            visualStackWindow,
            JScrollPane.VERTICAL_SCROLLBAR_ALWAYS,
            JScrollPane.HORIZONTAL_SCROLLBAR_NEVER
        );
        textField = textInput()
        pushButton = JButton("PUSH").apply{
           background = DefaultWindowsInterface.themeColorBG
           foreground = Color.WHITE
           cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
           font = this@StackWindowUsable.font
           addActionListener {
                if (textField.inputVerifier.verify(textField)) {
                    visualStackWindow.push(textField.text.toInt())
                    textField.text = ""
                    visualStackWindow.setCamCentered(scrollPane)
                }
           }
            size= Dimension(100,50)
        }
        popButton = JButton("POP").apply {
            cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
            font = this@StackWindowUsable.font
            foreground = Color.WHITE
            background = Color.BLACK
            addActionListener {
                visualStackWindow.pop();
                visualStackWindow.setCamCentered(scrollPane)
            }
            size= Dimension(200,50)
        }
        scrollPane.setBounds(0, 0, myWidth, myHeight)
        val mainPanel = JPanel().apply {
            layout= BorderLayout()
            val subPanel = JPanel(GridLayout(1, 2, 5, 5)).apply {
                add(textField)
                add(pushButton)
            }
           add(subPanel, BorderLayout.NORTH)
           add(popButton, BorderLayout.SOUTH)
           setBounds(myWidth / 2 - 150, myHeight - 200, 200, 100)
        }
        layeredPane.add(scrollPane)
        this.add(layeredPane, BorderLayout.CENTER)
        this.add(mainPanel, BorderLayout.SOUTH)
    }

    private fun textInput(): JTextField {
        return JTextField().apply {
            size= Dimension(100,50)
            cursor = Cursor.getPredefinedCursor(Cursor.TEXT_CURSOR)
            font = this@StackWindowUsable.font
            background = DefaultWindowsInterface.themeColorBG
            foreground = Color.WHITE
            horizontalAlignment = JTextField.CENTER
            toolTipText = "Value of the next block (Int): "
            inputVerifier = object : InputVerifier() {
                override fun verify(input: JComponent?): Boolean {
                    return try {
                        (input as? JTextField)?.text?.toInt()
                        true
                    } catch (_: NumberFormatException) {
                        false
                    }
                }
            }
            addKeyListener(object : KeyListener {
                override fun keyTyped(e: KeyEvent?) {
                    if (inputVerifier!!.verify(this@apply) && e?.keyChar == '\n') {
                        visualStackWindow.push(text.toInt())
                    }
                    if (e?.keyChar == 127.toChar()) {
                        visualStackWindow.pop()
                    }
                }

                //Discarded
                override fun keyPressed(e: KeyEvent?) {
                }

                override fun keyReleased(e: KeyEvent?) {
                }

            })
        }
    }
}

fun main() {
    SwingUtilities.invokeLater {
        JFrame("Stack Window").apply {
            add(StackWindowUsable())
            size = Dimension(DefaultWindowsInterface.width, DefaultWindowsInterface.height)
            defaultCloseOperation = JFrame.EXIT_ON_CLOSE
            isVisible = true
        }
    }
}